### PR TITLE
Deepspeed fix : save the normal model too

### PR DESCRIPTION
**Pull request by rom1504**
useful to be able to get the model for generation even when using deepspeed

### CODE REVIEW

**Comment by janEbert** [comment]
Hey, I actually had this implemented as well but removed it in c4223f936c2f5099c69a7f23be1bc087f4da2a29 as it was too confusing. The reason it's getting confusing is that these checkpoints become messed up once you start using ZeRO stage 3.

There is a [small hint on how to convert DeepSpeed checkpoints to standard checkpoints in the `save_model` function](https://github.com/lucidrains/DALLE-pytorch/blob/main/train_dalle.py#L358-L359). Perhaps it should be expanded?

**Comment by janEbert** [comment]
A possibility would be adding a check, like so:
```python
if deepspeed_config.get('zero_optimization', {}).get('stage', 0) >= 2:
    return
```

However, I think this again results in confusion due to the inconsistency: now you sometimes get standard checkpoints with DeepSpeed but sometimes you don't. What do you think?

**Comment by rom1504** [comment]
I think we should have a usable dalle.pt saved regularly in all cases.
It's useful for debugging and for generation. For
* the non distributed mode : that already works in main
* in deepspeed without zero : that starts working with this PR
* in deepspeed with zero : I think we should use something to make this happen, maybe you know how ?

So imo the change of this PR improves a bit the situation, even if it indeeds doesn't solve it for zero.

How can we make it work with zero ?

**Comment by afiaka87** [comment]
> I think we should have a usable dalle.pt saved regularly in all cases.
> It's useful for debugging and for generation. 

I agree with this - if the weights can be coalesced into a single file then they should be. I've personally had trouble using the given "convert_to_fp32.py" script that's given by ZeRO stages 1/2/3, but it _should_ work. 

@janEbert I'm all for usablity concerns - you're right, this stuff is pretty confusing. Having said that - this whole codebase is kinda tough to wrangle in the first place depending on which of the dozens of potential configurations you're trying to use e.g. horovod, deepspeed, deepspeed with apex, no-deepspeed but still using sparse attention, deepspeed _and_ using sparse attention, no-deepspeed but still using their FusedAdam on the GPU, etc. The README is like a mile long. 

Perhaps it's time to admit that the average user of this codebase should be comfortable with such complications?

At any rate - your solution works well enough, but I think more likely we should raise an exception with a friendly message saying something like:

```python
if deepspeed_config.get('zero_optimization', {}).get('stage', 0) >= 2:
    print(f"Checkpoints made with DeepSpeed ZeRO Stages 2 and 3 will be stored in {DS_CP_PATH}")
    print(f"As such, they will require DeepSpeed as a dependency in order to resume from or generate with.")
    print("See <insert deespeed conversion script link here> for details on how to convert your ZeRO stage 2/3 checkpoint to a single file.")
    print("If using a single GPU, consider running with apex automatic mixed precision instead for a similar speedup to ZeRO.")
    return
```

These things are complex - but as long as we're informing users of the complexity and sending them to the correct github issues to help themselves, then I think we are okay. 

@rom1504 the proper way to make it work with zero is to use a script given by them called `zero_to_fp32`.

Have you tried using DeepSpeed Zero stage 1/2/3 yet @rom1504? Even running with 32 bit instead of 16 bit I still get a dramatic speedup from ~8 samples/second to ~60 samples/second.

If we can't get DeepSpeed checkpoints to store in a single `dalle.pt`, the "deepspeed lock-in" way to do it would be to re-work `generate.py` in order for it to work with DeepSpeed's checkpoints. I think there is a page on deepspeed's docs about inference that @janEbert has suggested.

**Comment by rom1504** [comment]
Hi @afiaka87 , about deepspeed zero speeds I would be really interested to discuss this more in the discord. I just tried this yesterday and didn't observe any speed up at all with zero. In fact it was slower with zero 3.
I've been having success with increasing the batch size and using the batch accumulation option of deepspeed to decrease interconnect pressure in multi node setting. Now reaching 200 sample/s with a big model (8 nodes with 2 gpus)
I'd be interested to learn more about how you use zero, I think you're probably using better settings.

Regarding dalle.pt I think it would be great to try to automatically use this zero_to_fp32 script when in zero mode. Adapting generate script is also an option but imo it's going to make things complex.

**Comment by afiaka87** [comment]
@rom1504 I'm gonna be quite busy for the weekend but I'll actually agree with you on that point - I was mistakenly conflating ZeRO with deepspeed itself (and just having a bit of a lapse in memory).

Indeed - your experience matches mine - ZeRO doesn't seem to help nearly as much as the gradient accumulation provided by deepspeed. I've found that this configuration using apex for mixed precision is actually blazing fast, and doesn't require DeepSpeed. 

Keep in mind many of my experiments are running on my RTX2070 (8GiB of vram). So I can't really get past 16 depth/16 heads/512 dim on my setup.

Anyway, here's my current `deepspeed_config`:

```python
deepspeed_config = {
    'scheduler': {
        'type': 'WarmupDecayLR',
        'params': {
            'total_num_steps': int(len(ds) / BATCH_SIZE * EPOCHS),
            'warmup_min_lr': 0,
            'warmup_max_lr': LEARNING_RATE,
            'warmup_num_steps': int((len(ds) / BATCH_SIZE) * EPOCHS * 0.02), # 2 percent of all steps
        },
    },
    'optimizer': {
        'type': 'AdamW',
        'params': {
            'lr': LEARNING_RATE,
            'betas': [
                0.9,
                0.95
            ],
            'eps': 1e-8,
            'weight_decay': 4e-2,
        },
    },
    'sparse_gradients': True,
    'train_batch_size': BATCH_SIZE,
    'gradient_clipping': GRAD_CLIP_NORM,
    'amp': {
        'enabled': True,
        'opt_level': "O1",
    },
```

You can probably ignore the sparse gradients bit - I haven't seen much improvement one way or the other with that one. 

You'll also need to do two casts to float (instead of type `Long` that is normally needed) in the `train_dalle.py`:

```python
    if args.fp16:
            images = images.half()
    elif deepseed_config.get('apex', {}) is not None: # pseudocode, i just hardcoded them to float in my actual script. 
        images = images.float()

# ... further down the loop when reconstructions are generated
       image = dalle.generate_images(text[:1], filter_thres=0.9).float()  
```

This somehow saves me _more_ memory the deepspeed fp16 with zero stage 2/3 - although I haven't done rigorous benchmarking and am confident that the multi-gpu situation is different.

**Comment by rom1504** [comment]
@afiaka87 thanks that's useful to hear! I'll try out apex.

also it seems fp16 with amp also fixes the (otherwise broken) image generation with fp16 ?
would be useful to change that in main imo. It's pretty inconvenient that fp16 disable the image generation during training feature

**Comment by afiaka87** [comment]
> @afiaka87 thanks that's useful to hear! I'll try out apex.
> 
> also it seems fp16 with amp also fixes the (otherwise broken) image generation with fp16 ?
> would be useful to change that in main imo. It's pretty inconvenient that fp16 disable the image generation during training feature

Yeah I don't really use fp16 anymore now that ZeRO has 32 bit support across all stages. I never tried using automatic mixed precision with the fp16 flag actually - I assumed DeepSpeed would just override it if amp is set. Indeed, though - there aren't any generation issues with this method and it uses less memory than 32 bit precision for sure.

And yeah, that open bug for image generation needs some love, but unfortunately there's not a lot of motivation to fix it. Jan is rather busy currently but even then, Julich doesn't let you run W&B so they have less of a need to see the reconstructions until after a training session.

Anyway, DeepSpeed is a beast to wrangle with and there are so many other potential bottlenecks in the code - I ran a few benchmarks using DeepSpeed's "profiler" config - really cool stuff. Shows timing in ms of the forward and backward pass and also drills down on the timing of individual components like the attention types. Definitely something we should run through rigorously with the variety of hyperparameters.

**Comment by rom1504** [comment]
@afiaka87 I just tried your settings with a command like this ` deepspeed  --hostfile=deepspeed_host train_dalle.py --depth=32 --taming --image_text_folder ... --truncate_captions --wandb_name amazon_products --dalle_output_file_name amp --batch_size=64 --deepspeed` and in my case (using 8 nodes with 2 gpu ; so that's per gpu batch size 4), with amp I get:
* with amp and your code change / ds config : 12GB of vram used ; 30 sample/s
* with fp16 option without your code change 8.9GB of vram used ; 46 sample/s
so it seems at least in my case fp16 is better

I'm going to check if I can fix the generate method for fp16 flag

**Comment by rom1504** [comment]
So we got a bit side tracked here.
@janEbert what do you think we should do here ?
Should I add this if zero 0 ?

I think this PR is improving things when not using zero

**Comment by afiaka87** [comment]
> So we got a bit side tracked here.
> @janEbert what do you think we should do here ?
> Should I add this if zero 0 ?
> 
> I think this PR is improving things when not using zero

Yeah - let's just add a check for whether or not ZeRO is being used or not and print a message to the user pointing at this PR and/or an issue we create explaining the situation. I believe Jan is quite busy currently - so we can just move forward so long as @lucidrains thinks it's good to go.

**Comment by afiaka87** [comment]
> Perhaps it's time to admit that the average user of this codebase should be comfortable with such complications?
> 
> At any rate - your solution works well enough, but I think more likely we should print a friendly message saying something like:
> 
> ```python
> if deepspeed_config.get('zero_optimization', {}).get('stage', 0) >= 2:
>     print(f"Checkpoints made with DeepSpeed ZeRO Stages 2 and 3 will be stored in {DS_CP_PATH}")
>     print(f"As such, they will require DeepSpeed as a dependency in order to resume from or generate with.")
>     print("See <insert deespeed conversion script link here> for details on how to convert your ZeRO stage 2/3 checkpoint to a single file.")
>     print("If using a single GPU, consider running with apex automatic mixed precision instead for a similar speedup to ZeRO.")
>     return
> ```
> 
> These things are complex - but as long as we're informing users of the complexity and sending them to the correct github issues to help themselves, then I think we are okay.

@lucidrains @rom1504 This was more or less my solution, feel free to correct me on the finer details.

**Comment by janEbert** [comment]
> And yeah, that open bug for image generation needs some love, but unfortunately there's not a lot of motivation to fix it. Jan is rather busy currently

Oh there's definitely motivation! It's just so damn hard! :)
Believe me, I've invested a lot of time into not only trying different configurations but also diving deep into DeepSpeed source code. Still, I haven't found a fix yet. Appreciate you looking into it as well, though, @rom1504! Hope you can find what I missed.

> @janEbert what do you think we should do here ?
> Should I add this if zero 0 ?

I actually now think you're right; it's less confusing to save the checkpoints and warning the user when they don't get saved due to more advanced settings. So I'm for it! Even with the confusion of finding that the previously used checkpoints don't change their last modified date when you use the advanced settings, you get the best of both worlds. :)
ZeRO stage 1 checkpoints should also still work with this; only when stage 2 and upward is used do they start with their special parameter format that causes checkpoints to break.

With @afiaka87's current implementation (really nice warnings, by the way), we print the same warning each epoch, though; I suggest moving that out of the function so it's only printed at startup.

**Comment by rom1504** [comment]
hi @janEbert and @afiaka87 I made the suggested change. Could you please check if that's good for you?

If yes, could you please have a look too @lucidrains and let's merge this?

TLDR: this pull request makes it possible to save the dalle.pt model when deepspeed is used but without zero. If zero is enabled, it keeps the previous behavior of not saving dalle.pt. For not deep speed, it keeps the previous behavior of saving dalle.pt

**Comment by afiaka87** [comment]
Give me about two hours and I can run this locally to make sure it's working. I have independently verified that this _method_ works as I used it to share my automatic mixed precision checkpoint as a 'dalle.pt' for mega/robvanvolts colab. 

Always good to give things a run on more than one setup though. Will report any issues after running some errands.

File: train_dalle.py [hunk 1]
```diff
@@ -374,6 +381,11 @@ def save_model(path):
 
 # training
 
+# save the initial model
+# that makes sure saving is working, to avoid waiting
+# a long time and not getting any output
+save_model(DALLE_OUTPUT_FILE_NAME)
```
Comment (afiaka87) [inline train_dalle.py:387]: Nitpicking at this point -

How about this instead? 
```python
# Saves a checkpoint before training begins to fail early when mis-configured. 
# See https://github.com/lucidrains/DALLE-pytorch/wiki/DeepSpeed-Checkpoints
```

File: train_dalle.py [hunk 2]
```diff
@@ -320,6 +320,12 @@ def group_weight(model):
     },
 }
 
+if deepspeed_config.get('zero_optimization', {}).get('stage', 0) >= 2:
+    print(f"Checkpoints made with DeepSpeed ZeRO Stages 2 and 3 will be stored in deepspeed checkpoint folder")
+    print(f"As such, they will require DeepSpeed as a dependency in order to resume from or generate with.")
+    print("See the deespeed conversion script for details on how to convert your ZeRO stage 2/3 checkpoint to a single file.")
+    print("If using a single GPU, consider running with apex automatic mixed precision instead for a similar speedup to ZeRO.")
+
```
[the remark this answers is not in the document]
Comment (rom1504) [reply train_dalle.py:328]: agreed, done

File: train_dalle.py [hunk 3]
```diff
@@ -360,7 +366,8 @@ def save_model(path):
             ),
         }
         torch.save(save_obj, str(cp_dir / DEEPSPEED_CP_AUX_FILENAME))
-        return
+        if deepspeed_config.get('zero_optimization', {}).get('stage', 0) >= 2: # see explanation for this at the top of the file
```
[the remark this answers is not in the document]
Comment (rom1504) [reply train_dalle.py:369]: do you mean putting that in the comment ? if so, done

File: train_dalle.py (hunk 1 above)
Comment (rom1504) [reply train_dalle.py:388]: done

**Comment by rom1504** [comment]
thanks, I think I adressed your comments, tell me if that's good enough for you!

**Review by afiaka87 — APPROVED** [review]

**Comment by afiaka87** [comment]
Yeah looks good - up to @lucidrains now.

**Comment by lucidrains** [comment]
lgtm!
